nullaway/src/main/java/com/uber/nullaway/fixserialization/XMLUtil.java

```diff
@@ -0,0 +1,202 @@
+   */
+  public static <T> DefaultXMLValueProvider<T> getValueFromTag(
+      Document doc, String key, Class<T> klass) {
+    Node node = getNodeFromDocument(doc, key);
+    if (node == null) {
+      return new DefaultXMLValueProvider<>(null, klass);
+    }
+    return new DefaultXMLValueProvider<>(node, klass);
+  }
+
+  /**
+   * Writes the {@link FixSerializationConfig} in {@code XML} format.
+   *
+   * @param config Config file to write.
+   * @param path Path to write the config at.
+   */
+  public static void writeInXMLFormat(FixSerializationConfig config, String path) {
+    DocumentBuilderFactory docFactory = DocumentBuilderFactory.newInstance();
+    try {
+      DocumentBuilder docBuilder = docFactory.newDocumentBuilder();
+      Document doc = docBuilder.newDocument();
+
+      // Root
+      Element rootElement = doc.createElement("serialization");
+      doc.appendChild(rootElement);
+
+      // Suggest
+      Element suggestElement = doc.createElement("suggest");
+      suggestElement.setAttribute("active", String.valueOf(config.suggestEnabled));
+      suggestElement.setAttribute("enclosing", String.valueOf(config.suggestEnclosing));
+      rootElement.appendChild(suggestElement);
+
+      // Annotations
+      Element annots = doc.createElement("annotation");
+      Element nonnull = doc.createElement("nonnull");
+      nonnull.setTextContent(config.annotationConfig.getNonNull().fullName);
+      Element nullable = doc.createElement("nullable");
+      nullable.setTextContent(config.annotationConfig.getNullable().fullName);
+      annots.appendChild(nullable);
+      annots.appendChild(nonnull);
+      rootElement.appendChild(annots);
+
+      // Output dir
+      Element outputDir = doc.createElement("path");
+      outputDir.setTextContent(config.outputDirectory);
+      rootElement.appendChild(outputDir);
+
+      // Writings
+      TransformerFactory transformerFactory = TransformerFactory.newInstance();
+      Transformer transformer = transformerFactory.newTransformer();
+      DOMSource source = new DOMSource(doc);
+      StreamResult result = new StreamResult(new File(path));
+      transformer.transform(source, result);
+    } catch (Exception e) {
+      throw new RuntimeException("Error happened in writing config." + e);
+    }
+  }
+
+  /**
+   * Locates the node at the given key. Key can be nested in format key_1:key_2:...:key_n.
+   *
+   * @param doc Document object of XML.
+   * @param key Nested key.
+   * @return Node if exists with the hierarchy specified by the key, otherwise {@code null}.
+   */
+  private static Node getNodeFromDocument(Document doc, String key) {
+    if (doc == null || key == null || key.equals("")) {
+      return null;
+    }
+    ArrayList<String> keys = new ArrayList<>(Arrays.asList(key.split(":")));
+    Collections.reverse(keys);
+    NodeList candidates = doc.getElementsByTagName(keys.get(0));
+    for (int i = 0; i < candidates.getLength(); i++) {
```

Comment:
This is probably fine given how small the XML file is and how there should be only one element with a given tag, but in the general case it seems inefficient to enumerate all elements with a tag then check their path. Aren't there already XML APIs for this sort of thing? XPath or similar? (Definitely better ways to do this for JSON/Yaml files and I imagine XML is the same)

Reply:
76eb3ab7f02e81213cb0b7a31c2db8466b8ef966

Added `XPath` API, looks much cleaner :D , Thank you very much.